llama-rs/src/lib.rs

```diff
@@ -1211,21 +1241,21 @@ impl InferenceSession {
         // First, sample the next token, using the stored last_logits;
         let next_token = model.sample_top_p_top_k(self, params, rng);
 
-        // Update the last_n_tokens list
-        self.last_n_tokens.push_front(next_token);
+        // Update the tokens for this session
+        self.tokens.push(next_token);
 
         // Then, evaluate the network again to compute the new last_logits
         model.evaluate(self, params.n_threads, &[next_token]);
 
         // Return the next token
-        if next_token == 2 {
-            Ok(OutputToken::EndOfText)
-        } else {
-            Ok(OutputToken::Token(&vocab.mapping[next_token as usize]))
-        }
+        Ok(OutputToken::from_id(vocab, next_token))
     }
 
     // todo: see if we can reduce the arguments here somehow - consolidate model and vocab maybe?
+    /// Helper function to run inference with this session and the given model and vocabulary.
+    ///
+    /// Note that this will "play back" all existing tokens in the session. If this is not desired
+    /// behaviour, consider implementing your own inference loop to customize the behavior.
```

Comment:
Given that the contents of `inference_with_prompt` are not exactly trivial, I'd say this behavior should be controlled by a function parameter.

The most interesting use case for prompt caching is something like Bing, where they feed the model a giant prompt to the model (this one is cached, so you save a lot of processing time), plus the user query at the end. In that case, you don't want to print the contents of the "pre-prompt", just the contents of the prompt you're currently feeding the system. But other than that the `inference_with_prompt` function works perfectly fine, so it would be nice if users that want to do this weren't forced to reimplement their own loops.

Reply:
Mm, yeah, I'm not sure what the best course of action here is. `inference_with_prompt` is getting a little unwieldy because it's hard to accommodate everyone's use cases. If you think a boolean's the right call here, I'll add it.

Reply:
But I agree, let's make sure we don't add too much bloat and direct people to the lower level functions. But just for this, since the function used to work the other way, let's add the boolean just to make transition easier. I generally don't like breaking changes. Even though this code doesn't have lots of users, it might be annoying if we have to tell people they need to drop down to the lower level API now when they could use this just fine before.

Reply:
Fair enough, will do in a bit!